### PR TITLE
added templated normalize function

### PR DESCRIPTION
**Pull request by meiravgri**
**Describe the changes in the pull request**

added static templated normalize vector function in VecSimAbstract.

**Which issues this PR fixes**
1.Generelize normalize function

**Main objects this PR modified**
1. VecSimAbstract
2. removed normalize function from utils

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/vec_sim_index.h [hunk 1]
```diff
@@ -55,4 +57,37 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
...
+        memcpy(output, input_vector, dim * sizeof(DataType));
+    } else { // else we are normalizing inplace.
+        output = (DataType *)input_vector;
+    }
+    double sum = 0;
+    for (size_t i = 0; i < dim; i++) {
+        double to_add = (double)output[i] * (double)output[i];
+        // Protect from overflow.
+        assert(MAX_FP64 - sum >= to_add);
+        sum += to_add;
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_index.h:81]: Why the double add? Just check for overflow.
Check those links https://stackoverflow.com/questions/15655070/how-to-detect-double-precision-floating-point-overflow-and-underflow
https://en.cppreference.com/w/cpp/numeric/fenv/fetestexcept

File: src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp [hunk 2]
```diff
@@ -130,11 +132,11 @@ VecSimQueryResult_List HNSWIndex::rangeQuery(const void *queryBlob, float radius
     void *timeoutCtx = nullptr;
     this->last_mode = RANGE_QUERY;
 
-    float normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine
-    if (this->metric == VecSimMetric_Cosine) {
-        // TODO: need more generic when other types will be supported.
-        memcpy(normalized_blob, queryBlob, this->dim * sizeof(float));
-        float_vector_normalize(normalized_blob, this->dim);
+    bool is_cosine = this->metric == VecSimMetric_Cosine;
+    float normalized_blob[this->dim *
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp:136]: I don't understand the purpose of the code. Avoid stack-allocated array with multiplication?

File: src/VecSim/vec_sim_index.h [hunk 3]
```diff
@@ -55,4 +57,37 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
...
+static const double MAX_FP64 = std::numeric_limits<double>::max();
+
+template <typename DataType>
+void VecSimIndexAbstract::NormalizeVector(const void *input_vector, size_t dim, DataType *output) {
+
+    // if output is not NULL and doesn't point to the same memory area as input.
+    if (output && output != (DataType *)input_vector) {
+        memcpy(output, input_vector, dim * sizeof(DataType));
+    } else { // else we are normalizing inplace.
+        output = (DataType *)input_vector;
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_index.h:74]: You are breaking `const` guard here. If you need to copy, copy outside of the function. recall the function purpose is to normalize.

**Review by DvirDukhan — COMMENTED** [review]
Thanks
I think we need to add the `normalize` flow to the benchmark suite before we continue with this PR.

File: src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp (hunk 2 above)
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp:136]: yes, the normalized blob size will be 0 if not cosine

File: src/VecSim/vec_sim_index.h (hunk 1 above)
Comment (meiravgri) [reply src/VecSim/vec_sim_index.h:81]: I tried a few kinds of library functions that check overflow and I've found that they might not be good enough,
due to the floating point rounding mechanism which depends on the value:
https://stackoverflow.com/questions/37830178/why-no-overflow-when-using-double-variable
So assuming we have reached the maximum value of fp32, and we keep adding reasonable values, these flags won't be set...

The solution here is to check overflow before adding

File: src/VecSim/vec_sim.cpp [hunk 4]
```diff
@@ -125,9 +125,8 @@ extern "C" size_t VecSimIndex_EstimateElementSize(const VecSimParams *params) {
 }
 
 extern "C" void VecSim_Normalize(void *blob, size_t dim, VecSimType type) {
-    // TODO: need more generic
     assert(type == VecSimType_FLOAT32);
-    float_vector_normalize((float *)blob, dim);
+    VecSimIndexAbstract<float>::NormalizeVector((float *)blob, dim);
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim.cpp:129]: move the function back to vec_utils
check for the supported types

**Review by DvirDukhan — COMMENTED** [review]
check what happens with the stack pointer with respect to array size 0

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 5]
```diff
@@ -7,6 +7,7 @@
 #include "VecSim/algorithms/brute_force/brute_force_factory.h"
 #include "VecSim/spaces/spaces.h"
 #include "VecSim/query_result_struct.h"
+#include "VecSim/utils/vec_utils.h" // NormalizeVector
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force.h:10]: remove comment

File: src/VecSim/algorithms/brute_force/brute_force_multi.h [hunk 6]
```diff
@@ -3,6 +3,7 @@
 #include "brute_force.h"
 #include "bfm_batch_iterator.h"
 #include "VecSim/utils/updatable_heap.h"
+#include "VecSim/utils/vec_utils.h" // NormalizeVector
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force_multi.h:6]: remove comment

File: src/VecSim/algorithms/brute_force/brute_force_single.h [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 #include "brute_force.h"
 #include "bfs_batch_iterator.h"
+#include "VecSim/utils/vec_utils.h" // NormalizeVector
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force_single.h:5]: remove comment

File: src/VecSim/algorithms/brute_force/brute_force_multi.h [hunk 8]
```diff
@@ -57,12 +58,11 @@ class BruteForceIndex_Multi : public BruteForceIndex<DataType, DistType> {
 template <typename DataType, typename DistType>
 int BruteForceIndex_Multi<DataType, DistType>::addVector(const void *vector_data, labelType label) {
 
-    DataType normalized_data[this->dim]; // This will be use only if metric == VecSimMetric_Cosine
+    DataType normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine.
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force_multi.h:61]: why the rename?

Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force_multi.h:61]: why the rename?

File: src/VecSim/algorithms/brute_force/brute_force_single.h [hunk 9]
```diff
@@ -81,12 +82,11 @@ template <typename DataType, typename DistType>
 int BruteForceIndex_Single<DataType, DistType>::addVector(const void *vector_data,
                                                           labelType label) {
 
-    DataType normalized_data[this->dim]; // This will be use only if metric == VecSimMetric_Cosine
+    DataType normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force_single.h:85]: why the rename?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 10]
```diff
@@ -1143,12 +1143,11 @@ int HNSWIndex<DataType, DistType>::addVector(const void *vector_data, const labe
 
     idType cur_c;
 
-    DataType normalized_data[this->dim]; // This will be use only if metric == VecSimMetric_Cosine.
+    DataType normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnsw.h:1146]: why the rename?

File: src/VecSim/algorithms/brute_force/brute_force_single.h (hunk 9 above)
Comment (DvirDukhan) [inline src/VecSim/algorithms/brute_force/brute_force_single.h:85]: why the rename?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 11]
```diff
@@ -1380,12 +1379,11 @@ VecSimQueryResult_List HNSWIndex<DataType, DistType>::topKQuery(const void *quer
 
     void *timeoutCtx = nullptr;
 
-    DataType normalized_data[this->dim]; // This will be use only if metric == VecSimMetric_Cosine.
+    DataType normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine.
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnsw.h:1382]: why the rename?

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 10 above)
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnsw.h:1146]: why the rename?

File: src/VecSim/utils/vec_utils.h [hunk 12]
```diff
@@ -4,6 +4,7 @@
 #include "VecSim/vec_sim_common.h"
 #include <VecSim/query_results.h>
 #include <utility>
+#include <cmath> //sqrt
```
Comment (DvirDukhan) [inline src/VecSim/utils/vec_utils.h:7]: remove comment

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 11 above)
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnsw.h:1382]: why the rename?

File: src/VecSim/utils/vec_utils.h [hunk 13]
```diff
@@ -66,3 +65,19 @@ const char *VecSimMetric_ToString(VecSimMetric vecsimMetric);
 const char *VecSimSearchMode_ToString(VecSearchMode vecsimSearchMode);
 
 size_t VecSimType_sizeof(VecSimType vecsimType);
+
+template <typename DataType>
+void NormalizeVector(DataType *input_vector, size_t dim) {
```
Comment (DvirDukhan) [inline src/VecSim/utils/vec_utils.h:70]:
```suggestion
void normalizeVector(DataType *input_vector, size_t dim) {
```

File: tests/unit/test_common.cpp [hunk 14]
```diff
@@ -3,6 +3,10 @@
 #include "VecSim/utils/arr_cpp.h"
 #include "VecSim/utils/updatable_heap.h"
 
+#include <cstdlib> // rand, RAND_MAX
+#include <limits>  //numeric_limits
+#include <cmath>   // exp
```
Comment (DvirDukhan) [inline tests/unit/test_common.cpp:8]: remove comments

File: tests/unit/test_common.cpp [hunk 15]
```diff
@@ -243,3 +247,52 @@ TEST_F(CommonTest, Max_Updatable_Heap) {
     ASSERT_EQ(heap.size(), 0);
     ASSERT_TRUE(heap.empty());
 }
+
+TEST_F(CommonTest, VecSim_Normalize_FP32_Vector) {
+    const size_t dim = 1000;
+    float v[dim];
+
+    // generate random values - always generates the same values
```
Comment (DvirDukhan) [inline tests/unit/test_common.cpp:255]: please set the random seed to be constant to do so
check other tests to see how

File: tests/unit/test_common.cpp [hunk 16]
```diff
@@ -243,3 +247,52 @@ TEST_F(CommonTest, Max_Updatable_Heap) {
...
+    }
+
+    // Change some of the vector's values so that the sum of the squared vector's
+    // values will overflow for floats but not for doubles.
+    v[dim - 3] = exp(44);
+    v[dim - 2] = exp(44);
+    v[dim - 1] = exp(44);
+
+    // Normalize the vector
+    VecSim_Normalize(v, dim, VecSimType_FLOAT32);
```
Comment (DvirDukhan) [inline tests/unit/test_common.cpp:268]: Consider even doing a templated test
https://stackoverflow.com/questions/17079702/gtest-testing-template-class

File: tests/unit/test_common.cpp [hunk 17]
```diff
@@ -243,3 +247,52 @@ TEST_F(CommonTest, Max_Updatable_Heap) {
...
+
+    // Normalize the vector
+    VecSim_Normalize(v, dim, VecSimType_FLOAT64);
+
+    // Check that the normelized vector norm is 1.
+    double norm = 0;
+    for (size_t i = 0; i < dim; ++i) {
+        norm += v[i] * v[i];
+    }
+    ASSERT_LE(1.0 * 0.99, norm);
```
Comment (DvirDukhan) [inline tests/unit/test_common.cpp:297]: ?

**Review by DvirDukhan — CHANGES_REQUESTED** [review]

File: src/VecSim/algorithms/brute_force/brute_force_multi.h (hunk 8 above)
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force_multi.h:61]: it was either normalized_blob or normalized_data so I chose one and made it consistent across the library

File: tests/unit/test_common.cpp [hunk 18]
```diff
@@ -248,14 +250,25 @@ TEST_F(CommonTest, Max_Updatable_Heap) {
     ASSERT_TRUE(heap.empty());
 }
 
-TEST_F(CommonTest, VecSim_Normalize_FP32_Vector) {
+template <typename DataType>
+class UtilsTests : public ::testing::Test {
+public:
```
Comment (DvirDukhan) [inline tests/unit/test_common.cpp:255]: remove line?

**Review by DvirDukhan — APPROVED** [review]

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**